### PR TITLE
[5.7] Fix issue that prevents manipulation of overloaded Eloquent model JSON columns

### PR DESCRIPTION
**Pull request by judahnator**
Hello Laravel Team!

The Problem:
-----------------

In various projects I have worked on I have seen this error when working with JSON columns:
```php
$myModel = MyModel::find($n);
$myModel->json_column->some_key = 'some value';
// PHP Notice:  Indirect modification of overloaded property App/MyModel::$json_column has no effect in...
```

The work around is to use the Laravel getter/setter to update the entire field at once and not just the bit that you want, like this:
```php
$myModel = MyModel::find($n);
$jsonTemp = $myModel->json_column;
$jsonTemp->some_key = 'some value';
$myModel->json_column = $jsonTemp;
unset($jsonTemp);
```

The Solution:
-----------------

This issue was discussed a bit in this [laravel/ideas](https://github.com/laravel/ideas/issues/1275) issue. If implemented, it would allow in-place changes of JSON column nested attributes. For example, lets say your JSON column had this data:
```json
{
  "an_array": [
    "foo",
    {
      "nested": "value"
    }
  ]
}
```

Using these changes, you could do this and it would work:
```php
$myModel = MyModel::find($n);
$myModel->json_column->an_array[1]->nested = 'edited value';
$myModel->save();
```

Changes Required For The User:
------------------------------------------

As you can see looking through the updated code I left existing functionality in place and added two new casts. "json_object" and "json_array". They are functionally identical, with the only change being if the initial value the user sees when creating a fresh object will be an object or array.

In order for the user to take advantage of these changes they will just need to change the cast they are using from "json" to "json_object" or "json_array".
```php
class MyModel extends Model
{
  $casts = [
    'json_column' => 'json_object'
  ]
}
```

Summary:
-------------

With these changes you can manipulate JSON attributes, even deeply nested ones, as if they were native PHP objects or arrays.

I believe I covered everything in tests. If I missed anything let me know.

I look forward to the feedback!

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
         "doctrine/inflector": "^1.1",
         "dragonmantank/cron-expression": "^2.0",
         "erusev/parsedown": "^1.7",
+        "judahnator/json-manipulator": "^1.2",
```
Comment (mfn) [inline composer.json]: `illuminate/database` can also used standalone, you would also need to add it to `Illuminate/Database/composer.json`.

Comment (judahnator) [reply composer.json]: Ah! Good catch! I appreciate the feedback and have updated the PR.

**Comment by taylorotwell** [comment]
No plans to add this to the core. Your library's first sentence says: "The short version is its a super memory heavy library".

**Comment by judahnator** [comment]
I appreciate the feedback and will use it to become a better programmer.

Y'all have a fantastic rest of the day!

**Comment by jordanade** [comment]
The underlying library has been updated to be more memory efficient. Any chance to take another look at this?

**Pull request closed without merging**